chia-keys/src/secret_key.rs

```diff
@@ -0,0 +1,302 @@
+        assert!(seed.len() >= 32);
+
+        const SALT: &[u8] = b"BLS-SIG-KEYGEN-SALT-";
+        let mut extracter = HkdfExtract::<sha2::Sha256>::new(Some(SALT));
+        extracter.input_ikm(seed);
+        extracter.input_ikm(&[0_u8]);
+        let h = extracter.finalize().1;
+
+        // TODO: = uninitialized
+        let mut sk = [0_u8; 48];
+        h.expand(&[0_u8, 48], &mut sk)
+            .expect("failed to generate secret key");
+        SecretKey(Scalar::from_okm(&sk))
+    }
+
+    pub fn from_bytes(b: &[u8; 32]) -> Option<SecretKey> {
+        let t = [
+            b[31], b[30], b[29], b[28], b[27], b[26], b[25], b[24], b[23], b[22], b[21], b[20],
+            b[19], b[18], b[17], b[16], b[15], b[14], b[13], b[12], b[11], b[10], b[9], b[8], b[7],
+            b[6], b[5], b[4], b[3], b[2], b[1], b[0],
+        ];
+        Scalar::from_bytes(&t).map(SecretKey).into()
+    }
+
+    pub fn to_bytes(&self) -> [u8; 32] {
+        let mut bytes = self.0.to_bytes();
+        bytes.reverse();
+        bytes
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(G1Projective::generator() * self.0)
+    }
+
+    fn to_lamport_pk(&self, idx: u32) -> [u8; 32] {
+        let ikm = self.to_bytes();
+        let not_ikm = flip_bits(ikm);
+        let salt = idx.to_be_bytes();
+
+        let mut lamport0 = ikm_to_lamport_sk(&ikm, &salt);
+        let mut lamport1 = ikm_to_lamport_sk(&not_ikm, &salt);
+
+        for i in (0..32 * 255).step_by(32) {
+            let hash = sha256(&lamport0[i..i + 32]);
+            lamport0[i..i + 32].copy_from_slice(&hash);
+        }
+        for i in (0..32 * 255).step_by(32) {
+            let hash = sha256(&lamport1[i..i + 32]);
+            lamport1[i..i + 32].copy_from_slice(&hash);
+        }
+
+        let mut hasher = Sha256::new();
+        hasher.update(lamport0);
+        hasher.update(lamport1);
+        hasher.finalize().try_into().unwrap()
+    }
+
+    pub fn derive_hardened(&self, idx: u32) -> SecretKey {
+        // described here:
+        // https://eips.ethereum.org/EIPS/eip-2333#derive_child_sk
+        SecretKey::from_seed(&self.to_lamport_pk(idx))
+    }
+}
+
+impl DerivableKey for SecretKey {
+    fn derive_unhardened(&self, idx: u32) -> Self {
+        let pk = self.public_key();
+
+        let mut hasher = Sha256::new();
+        hasher.update(pk.to_bytes());
+        hasher.update(idx.to_be_bytes());
+        let digest = hasher.finalize();
+
+        // in an ideal world, we would not need to reach for the sledge-hammer of
+        // num-bigint here. This would most likely be faster if implemented in
+        // Scalar directly.
+
+        // interpret the hash as an unsigned big-endian number
+        let mut scalar = BigUint::from_bytes_be(digest.as_slice());
+
+        let q = BigUint::from_bytes_be(&[
+            0x73, 0xed, 0xa7, 0x53, 0x29, 0x9d, 0x7d, 0x48, 0x33, 0x39, 0xd8, 0x08, 0x09, 0xa1,
```

Comment:
Refactor into a `group_order.rs` file or something. We really should consider forking bls12_381 and moving this into there.

Hmm, actually, might be worth trying the already existing `from_u512` method [here](https://github.com/mikelodder7/bls12_381_plus/blame/main/src/scalar.rs#L326). Even though it has a lot of padding, it uses some clever techniques that might made it faster than classical mod. (We could also just implement a `from_u256` method.)

Reply:
(Although I get that at this point, you're just trying to get something that works even if non-optimized.)

Reply:
yeah, I hope to revisit this in the future. I think I've spent more than its fair share to get this to work at all.